app/controllers/admin/orders_controller.rb

```diff
@@ -0,0 +1,121 @@
+class Admin::OrdersController < Admin::AdminController
+  include Pagy::Backend
+  before_action :find_order, only: [:show, :edit, :update, :cancel]
+
+  def index
+    @pagy, @orders = pagy(filtered_orders.order(sort_column => sort_direction))
+  end
+
+  def show
+    @order = Order.includes(order_items: :product).find(params[:id])
+  end
+
+  def edit; end
+
+  def update
+    ActiveRecord::Base.transaction do
+      handle_cancel_order if cancel_reason_present?
+
+      if update_order && update_address
+        flash[:success] = t("admin.orders_admin.update.success")
+        redirect_to admin_order_path(@order)
+      else
+        handle_update_error
+      end
+    end
+  end
+
+  def cancel
+    @order = Order.find_by(id: params[:id])
+
+    if @order.nil?
+      redirect_to admin_orders_path,
+                  alert: t("admin.orders_admin.cancel.not_found")
+    elsif @order.cancelled!
+      redirect_to admin_orders_path,
+                  notice: t("admin.orders_admin.cancel.success")
+    else
+      redirect_to admin_orders_path,
+                  alert: t("admin.orders_admin.cancel.failure")
+    end
+  end
+
+  private
+
+  def find_order
+    @order = Order.find_by(id: params[:id])
+    return if @order
+
+    flash[:alert] = t ".not_found"
+    redirect_to root_path
+  end
+
+  def order_params
+    params.require(:order).permit Order::UPDATE_ORDER_ADMIN
+  end
+
+  def address_params
+    params.require(:order).require(:address).permit(
+      Address::ADDRESS_REQUIRE_ATTRIBUTES
+    )
+  end
+
+  def cancel_reason_present?
+    params[:order][:cancel_reason].present?
+  end
+
+  def handle_cancel_order
+    @order.cancel_reason = params[:order][:cancel_reason]
+    @order.status = "cancelled"
+    restore_product_stock
+  end
+
+  def restore_product_stock
+    @order.order_items.each do |order_item|
+      product = order_item.product
+      product.stock += order_item.quantity
+      product.save
+    end
+  end
+
+  def update_order
+    @order.update(order_params)
+  end
+
+  def update_address
+    @order.address.update(address_params)
+  end
+
+  def handle_update_error
+    flash[:alert] = if @order.errors.any?
+                      t(".update.error_with_order",
+                        errors: order_errors_message)
+                    else
+                      t(".update.error_with_address")
+                    end
+    redirect_to edit_admin_order_path(@order), status: :unprocessable_entity
+  end
+
+  def order_errors_message
+    @order.errors.full_messages.join(", ")
+  end
+
+  def filtered_orders
+    orders = Order.all
+    orders = orders.where(status: params[:status]) if params[:status].present?
```

Comment:
```suggestion
    orders = Order.by_status(params[:status])
```
scope :by_status, -> (status){where(status: status) if status.present?}